client/asset/dcr/dcr.go

```diff
@@ -4527,3 +4546,178 @@ func float64PtrStr(v *float64) string {
+			dcr.mempoolRedeems[secretHash] = &mempoolRedeem{txHash: *txHash, firstSeen: time.Now()}
+			dcr.mempoolRedeemsMtx.Unlock()
+		}
+		if !abandonTx {
+			if tx.Confirmations >= reqiredRedeemConfirms {
+				deleteMempoolRedeem = true
+			}
+			return &asset.ConfirmRedemptionStatus{
+				Confs:  uint64(tx.Confirmations),
+				Req:    reqiredRedeemConfirms,
+				CoinID: coinID,
+			}, nil
+		}
+	}
+
+	if !abandonTx && !errors.Is(err, asset.CoinNotFoundError) {
+		return nil, fmt.Errorf("problem searching for redemption transaction %s: %w", txHash, err)
+	}
+
+	// Redemption transaction is missing from the point of view of our node!
+	// Unlikely, but possible it was redeemed by another transaction. We
+	// assume a contract past its locktime cannot make it here, so it must
+	// not be refunded. Check if the contract is still an unspent output.
+
+	txHash, vout, err := decodeCoinID(redemption.Spends.Coin.ID())
+	if err != nil {
+		return nil, err
+	}
+
+	var spent bool
+	_, err = dcr.wallet.UnspentOutput(dcr.ctx, txHash, vout, 0)
+	if err != nil {
+		if errors.Is(err, asset.CoinNotFoundError) {
+			spent = true
+		} else {
+			return nil, err
+		}
+	}
+
+	if spent {
+		// First find the block containing the output itself.
+		scriptAddr, err := stdaddr.NewAddressScriptHashV0(redemption.Spends.Contract, dcr.chainParams)
+		if err != nil {
+			return nil, fmt.Errorf("error encoding contract address: %w", err)
+		}
+		_, pkScript := scriptAddr.PaymentScript()
+		outFound, block, err := dcr.externalTxOutput(dcr.ctx, newOutPoint(txHash, vout),
+			pkScript, time.Now().Add(-60*24*time.Hour)) // search up to 60 days ago
+		if err != nil {
+			return nil, err // possibly the contract is still in mempool
+		}
+		// Try to find a transaction that spends it.
+		spent, err := dcr.isOutputSpent(dcr.ctx, outFound)
+		if err != nil {
+			return nil, fmt.Errorf("error checking if contract %v:%d is spent: %w", *txHash, vout, err)
+		}
+		if spent {
+			vin := -1
+			spendTx := outFound.spenderTx
+			for i := range spendTx.TxIn {
+				sigScript := spendTx.TxIn[i].SignatureScript
+				sigScriptLen := len(sigScript)
+				if sigScriptLen < dexdcr.SwapContractSize {
+					continue
+				}
+				// The spent contract is at the end of the signature
+				// script. Lop off the front half.
+				script := sigScript[sigScriptLen-dexdcr.SwapContractSize:]
+				_, _, _, sh, err := dexdcr.ExtractSwapDetails(script, dcr.chainParams)
+				if err != nil {
+					// This is not our script, but not
+					// necessarily a problem. Ignoring the error.
+					continue
+				}
+				if bytes.Equal(sh[:], secretHash[:]) {
+					vin = i
+					break
+				}
+			}
+			if vin >= 0 {
+				_, height, err := dcr.wallet.GetBestBlock(dcr.ctx)
+				if err != nil {
+					return nil, err
+				}
+				hash := spendTx.TxHash()
+				return &asset.ConfirmRedemptionStatus{
+					Confs:  uint64(height - block.height),
+					Req:    reqiredRedeemConfirms,
+					CoinID: toCoinID(&hash, uint32(vin)),
+				}, nil
```

Comment:
If `Confs` is `< reqiredRedeemConfirms` should we add this redemption to `mempoolRedeems` so we can avoid future calls to `externalTxOutput`?

Reply:
Good idea, done.